src/math/power.c

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+//
+//
+
+ /* Include Files */
+#include <sof/math/power.h>
+#include <sof/debug/panic.h>
+#include <sof/audio/format.h>
+/* define constant */
+#define POWER_B_NUMERATOR 0x4000000000000ULL
+#define POWER_MAX_LIMIT 0x8000
+
+/* Function Definitions */
+/**
+ * Arguments	: int64_t denominator
+ *
+ * Return Type	: int32_t
+ */
+static inline int32_t div_s32s64(int64_t denominator)
+{
+	uint64_t b_denominator;
+	int32_t quotient;
+
+	if (denominator == 0) {
+		assert(denominator);
+	} else {
+		b_denominator =
+		    (denominator < 0) ? ~(uint64_t)denominator + 1 : (uint64_t)denominator;
+		b_denominator = POWER_B_NUMERATOR / b_denominator;
+
+		if (denominator < 0)
+			/* for negative denominator, typecast uint64_t to int32_t for */
+			/* negative quotient */
+			quotient = (int32_t)-(int64_t)b_denominator;
+		else
+			quotient = (int32_t)b_denominator;
+	}
+	return quotient;
+}
+
+/**
+ * Arguments	: int32_t b
+ * base input range [-32 to +32]
+ *
+ *		: int32_t e
+ * exponent input range [ -3 to +3 ]
+ *
+ * Return Type	: int32_t
+ * power output range [-32768 to + 32768]
+ *
+ * +------------------+------------------+------------------+--------+--------+--------+
+ * | base	      | exponent	 | power(returntype)|	b    |	e     |	 p     |
+ * +----+-----+-------+----+-----+-------+----+----+--------+--------+--------+--------+
+ * |WLen| FLen|Signbit|WLen| FLen|Signbit|WLen|FLen|Signbit | Qformat| Qformat| Qformat|
+ * +----+-----+-------+----+-----+-------+----+----+--------+--------+--------+--------+
+ * | 32 | 25  |	 1    | 32 | 29	 |  1	 | 32 | 15 |   1    | 6.26   | 2.30   | 16.16  |
+ * +------------------+------------------+------------------+--------+--------+--------+
+ */
+int32_t power_int32(int32_t b, int32_t e)
+{
+	int32_t i;
+	int32_t k;
+	int32_t multiplier;
+	int32_t p;
+
+	p = POWER_MAX_LIMIT;
+
+	if (e < 0) {
+		e = -e;
+		multiplier = (b == 0) ? INT32_MAX : div_s32s64((int64_t)b);
```

Comment:
I've sent to you offline an optimization idea that seemed to work for this code line. Please check if's OK.

Reply:
Thanks - It works well .